scripts/research/b2_03_impulse_morphology_lib.py

```
@@ -0,0 +1,1419 @@
+    reference from `events` alone would silently drop any hourly T whose
+    PRE_VOL_60 is valid but every W's D_W(T) is exactly zero -- conflating
+    event-admission population with decision-reference population.
+    """
+    out = [dict(event) for event in events]
+    if not out:
+        return out
+
+    vol_times, vol_values = construct_hourly_vol_grid(frame)
+    vol_states, vol_scores = _causal_tertile_states(vol_values, vol_times)
+    vol_state_by_t = {
+        int(t): int(state) for t, state in zip(vol_times, vol_states)
```

Ruff (0.16.3)

[warning] `zip()` without an explicit `strict=` parameter

Add explicit value for parameter `strict=`

(B905)